Explain how to install and configure curator (#952)
Scalingo/documentation: _posts/platform/getting-started/2000-01-01-getting-started-with-elk.md

}
```

## Curator

### Installing Curator

> Keeping all your application logs is not always feasible due to the huge amount of storage that would be needed. However logs are usually only relevant for a short period of time. That's why it's current practice to destroy documents if they get too old.

Reviewer: ```suggestion
Keeping all your application logs is not always feasible due to the huge amount of storage that would be needed. However logs are usually only relevant for a short period of time. That's why the current best practice is to destroy logs after a certain amount of time.
```

> This is where [Curator](https://www.elastic.co/guide/en/elasticsearch/client/curator/5.8/index.html) is needed. This project is designed to let you managed your indices life cycle.

Reviewer: ```suggestion
This is where [Curator](https://www.elastic.co/guide/en/elasticsearch/client/curator/5.8/index.html) is needed. This project is designed to let you manage your indices life cycle.
```

Curator is written in Python. In order to install its dependencies, you can modify the `.buildpacks` file to also contain the Python buildpack.
The `.buildpacks` file should have the following content:

```
https://github.com/Scalingo/buildpack-jvm-common.git
https://github.com/Scalingo/python-buildpack.git
https://github.com/Scalingo/logstash-buildpack.git
```

In order to instruct the Python buildpack to install Curator and its dependencies, create a file named `requirements.txt` at the root of your project:

```
PyYAML==5.3.1
elasticsearch-curator==5.8.0
```

> Finally you'll need a file to write a script that will run curator every 12h. This can be achieved via a simple bash script.

Reviewer: ```suggestion
Finally you need to run Curator every 12h. This can be achieved via a shell script:
```

Create a file named `curator.sh` with the following content:

```bash
#!/bin/bash
set -ex

while true; do
  echo "Running Curator"
  curator --config curator.yml log-clean.yml
  # Run Curator every 12h
  sleep 43200
done
```

> Finally you need to tell Scalingo to start Curator.

Reviewer: ```suggestion
Finally you need to instruct Scalingo to start Curator.
```

This can be achieved by adding the following line to your `Procfile`:
```yaml
curator: ./curator.sh
```

### Configuration

All the hard stuff is now done.

Next step is to configure Curator. First you need to configure how Curator connects to your database:

```
---
client:
  hosts:
    - ${ELASTICSEARCH_HOST}
  http_auth: ${ELASTICSEARCH_AUTH}
logging:
  loglevel: INFO
  logfile:
  logformat: default
```

> Curator cannot use the `ELASTICSEARCH_URL`. You need to define 2 other environment variables on your app duplicating `ELASTICSEARCH_URL` content.

Reviewer: ```suggestion
Curator cannot use the `ELASTICSEARCH_URL` environment variable. You need to define two other environment variables on your app, duplicating `ELASTICSEARCH_URL` content.
```

> Hence if your `ELASTICSEARCH_URL` variable is set to `http://user:password@host:port`, you need to define 2 environment variables:

Reviewer: ```suggestion
Hence if your `ELASTICSEARCH_URL` variable is set to `http://user:password@host:port`, you need to define two environment variables:
```

```
ELASTICSEARCH_HOST=host:port
ELASTICSEARCH_AUTH=user:password
```

The last step is to configure your indices life cycle. This is based on your indices names.
When you configured Logstash, we asked you to configure your index with the following name `sc-apps-%{+YYYY.MM.dd}`.
We will configure Curator to parse the date contained on those names and delete the ones that are too old.

```yaml
actions:
  1:
    action: delete_indices
    description: Delete old log indices
    options:
      ignore_empty_list: True
      disable_action: False
    filters:
    - filtertype: pattern
      kind: prefix
      value: ${LOGS_INDICES_PREFIX}
    - filtertype: age
      source: name
      direction: older
      timestring: '%Y.%m.%d'
      unit: days
      unit_count: ${LOGS_RETENTION_DAYS}
```

> You now need to add two environment variables:

Reviewer: It's kinda odd to read "two environment variables" but actually only seeing one. I would rephrase to prevent this odd thing. Moreover you don't explain what this environment variable should contain (even though the variable name is almost self-explanatory).

Suggestion:
```suggestion
You now need to add two environment variables. The first environment variable is `LOGS_INDICES_PREFIX`. It configures the index pattern that should be affected by this policy. Setting this variable to `sc-apps` prevent Curator from deleting the other indices that are stored in the same database.

The second environment variable is `LOGS_RETENTION_DAYS`. It configures the retention time of your logs (in days). Setting this variable to `10`, Curator will delete an index if it is 10+ days old.

```
LOGS_INDICES_PREFIX=sc-apps 
LOGS_RETENTION_DAYS=10 
```
```

If this suggestion suits you, you can remove the text below.

```
PyYAML==5.3.1
elasticsearch-curator==5.8.0
```

Finally you'll need a file to write a script that will run curator every 12h. This can be achieved via a simple bash script.
Create a file named `curator.sh` with the following content:

```
#!/bin/bash
set -ex

while true; do
  echo "Running curator"
  curator --config curator.yml log-clean.yml
  # Run curator every 12h
  sleep 43200
done
```

Finally you need to tell Scalingo to start curator.
This can be archived via by adding the following line to your `Procfile`:
```
curator: ./curator.sh
```

### Configuration

All the hard stuff is now done.

Next step is configuring curator.
First you'll need to configure how curator will connect to your database.

```
---
client:
  hosts:
    - ${ELASTICSEARCH_HOST}
  http_auth: ${ELASTICSEARCH_AUTH}
logging:
  loglevel: INFO
  logfile:
  logformat: default
```

Sadly curator cannot use the `ELASTICSEARCH_URL` you will need to define 2 other environnement variable on your app:

So if your `ELASTICSEARCH_URL` is set to `http://user:password@host:port`, you'll need to define 2 environnement variables:

```

> ELASTICSEARCH_HOST=host:port

Author: They need to add use_ssl and other cert stuff. Not sure we want to add this to this page. It's already pretty long...

ELASTICSEARCH_AUTH=user:password
```

The last step is to configure your indices life cycle. This is based on your index names.
When you configured logstash, we asked you to configure your index with the following name `sc-apps-%{+YYYY.MM.dd}`.
We will configure Curator to parse the date contained on those names and delete the ones that are too old.

> ```

Reviewer: ```suggestion
```yaml
```

actions:
  1:
    action: delete_indices
    description: Delete old log indices
    options:
      ignore_empty_list: True
      disable_action: False
    filters:
    - filtertype: pattern
      kind: prefix
      value: ${LOGS_INDICES_PREFIX}
    - filtertype: age
      source: name
      direction: older
      timestring: '%Y.%m.%d'
      unit: days
      unit_count: ${LOGS_RETENTION_DAYS}
```

> You will now need to add two environnement variables:

Reviewer: ```suggestion
You now need to add two environment variables:
```

```
LOGS_INDICES_PREFIX=sc-apps
```

> This variable let you filter index that should be affected by this policy (so you wont delete other index that are stored in the same database);

Reviewer: ```suggestion
This variable lets you filter index that should be affected by this policy (so you wont delete other indices that are stored in the same database):
```

The parenthesis does not look relevant. Isn't it just a sentence?

```
LOGS_RETENTION_DAYS=10
```

> This variable let you configure the retention time of your logs. Here it will delete an index if it is 10+ days old.

Reviewer: ```suggestion
This variable lets you configure the retention time of your logs. With this configuration, Curator will delete an index if it is 10+ days old.
```

}
```

## Curator

### Installing Curator

Since logs are only relevant for a short period of time, it is current
practice to remove logs that are too old to be relevant. This is done to
reduce the load on the database and limit the disk usage.

This is where
[Curator](https://www.elastic.co/guide/en/elasticsearch/client/curator/5.8/index.html)
is needed. This project is designed to let you manage your indices life cycle.

Curator is written in Python. In order to install its dependencies, you
can modify the `.buildpacks` file to also contain the Python buildpack. The
`.buildpacks` file should have the following content:

```
https://github.com/Scalingo/buildpack-jvm-common.git
https://github.com/Scalingo/python-buildpack.git
https://github.com/Scalingo/logstash-buildpack.git
```

In order to instruct the Python buildpack to install Curator and its
dependencies, create a file named `requirements.txt` at the root of your
project:

```
PyYAML==5.3.1
elasticsearch-curator==5.8.0
```

Curator is not a deamon, it is designed as a one-off process. To be able

> to run it on Scalingo you'll need to write a simple bash script that launch

Reviewer: Nothing is "simple" :smiling_imp: 

```suggestion
to run it on Scalingo you need to write a Bash script that executes
```

> Curator regularly. This can be achieved via a simple shell script.

Reviewer: ```suggestion
Curator regularly.
```

You already wrote that

```bash
#!/bin/bash
set -ex

while true; do
  echo "Running Curator"
  curator --config curator.yml log-clean.yml
  # Run Curator every 12h
  sleep 43200
done
```

Finally you need to instruct Scalingo to start Curator.
This can be achieved by adding the following line to your `Procfile`:

```yaml
curator: ./curator.sh
```

### Configuration

All the hard stuff is now done.

Next step is to configure Curator. First you need to configure how Curator
connects to your database:

```
---
client:
  hosts:
    - ${ELASTICSEARCH_HOST}
  http_auth: ${ELASTICSEARCH_AUTH}
logging:
  loglevel: INFO
  logfile:
  logformat: default
```

Curator cannot use the `ELASTICSEARCH_URL` environment variable. You
need to define two other environment variables on your app, duplicating
`ELASTICSEARCH_URL` content.
Hence if your `ELASTICSEARCH_URL` variable is set to
`http://user:password@host:port`, you need to define 2 environment variables:

```
ELASTICSEARCH_HOST=host:port
ELASTICSEARCH_AUTH=user:password
```

The last step is to configure your indices life cycle. This is based on your
indices names.
When you configured Logstash, we asked you to configure your index with the

> following name `sc-apps-%{+YYYY.MM.dd}`.

Reviewer: I think this sentence is not correctly placed. We don't really understand why you remind us about it here. It is actually related to the value of `LOGS_INDICES_PREFIX` you explain a few lines later. Maybe move this sentence later?

Reviewer: You didn't addressed my question here I think

Author: Or just remove it.

> Create a file named: `log-clean.yml`. This configuration parse the indices

Reviewer: ```suggestion
Create a file named `log-clean.yml`. This configuration parses the indices
```

> names stored in elasticsearch and remove the one that too old.

Reviewer: ```suggestion
names stored in Elasticsearch and removes the ones that are too old.
```
